### DISCUSSION TITLE
Capturing time

### DISCUSSION THREAD

**Discussion by jamesmckechnie** [UKHomeOffice/design-system#334] [Ideas]
We are looking into better ways that we can capture time on our service. On COP we have two use cases; to capture:

- a time of day(e.g 1:34pm)
- a length of time (e.g. 1 hour 34 minutes)

We currently have a couple flavours of designs around COP and are looking to get a consistent on nailed down. 

On [a Figma page](https://www.figma.com/file/KhtVXkFraKl8kyGP1HuFzj/Capturing-time?node-id=0%3A1), I have collated a few examples found across Home Office and on other sources like the [GDS GitHub backlog](https://github.com/alphagov/govuk-design-system-backlog/issues/173). 

We also talked briefly about this on a [Slack thread](https://homeofficedigital.slack.com/archives/C0FR8PE8Z/p1645095726083499).

**Comment by jamesmckechnie** [comment]
### Multiple inputs
**COP - Events and the border**
![image](https://user-images.githubusercontent.com/34271054/154699144-e797809d-2b81-4a3b-8757-08557e0f01d0.png)

**Asylum ops - appointment booking**
![image](https://user-images.githubusercontent.com/34271054/154699290-6e3117c1-7959-4ecb-9bbc-58eb6f1c66b6.png)

**PPPT - LEDS**
![image](https://user-images.githubusercontent.com/34271054/154699359-a7b244f7-97bc-4753-ba1d-41c5814d2430.png)

### Single input
**COP - OAB**
![image](https://user-images.githubusercontent.com/34271054/154699666-6403b415-e934-4f17-bb00-83f074ca8faf.png)

**PPPT - MAAPPS**
![image](https://user-images.githubusercontent.com/34271054/154699768-94944353-6d84-4278-b4a0-85e7eddc93e8.png)

### Multiple inputs with AM/PM selector
**[GDS backlog - HMRC](https://github.com/alphagov/govuk-design-system-backlog/issues/173#issuecomment-439047514)**

### Autocomplete with dropdown suggestions
[**GDS backlog - Content publisher**](https://github.com/alphagov/govuk-design-system-backlog/issues/173#issuecomment-500375614)

**Comment by isobelseacombe** [comment]
Looks like there's an issue in the backlog that is looking into capturing date and time in code: #230

**Comment by isobelseacombe** [comment]
Also, this issue has been raised to show an example of displaying time in Callisto https://github.com/UKHomeOffice/design-system/issues/335

**Comment by Sarah-Lombardi** [comment]
In Callisto we add start time and finish time. That counts up their hours and attributes those hours to things they agreed in their contract. It automatically inserts a colon into the time (ie 08:00 not 8 or 8.00) if the user misses it.

When users view what they've entered it appears as a range with 'to' between the times eg 08:00 to 18:00.

For shifts, there is also a 'This is a X hour shift' note to confirm what will be taken off their total.

**Comment by Sarah-Lombardi** [comment]
<img width="622" alt="Screenshot 2022-02-18 at 14 29 18" src="https://user-images.githubusercontent.com/39190312/154705548-202e7122-a947-402f-868e-dc7b74f11c11.png">
<img width="554" alt="Screenshot 2022-02-18 at 14 28 55" src="https://user-images.githubusercontent.com/39190312/154705557-04764038-9512-44be-8576-e7cc918c5512.png">
<img width="615" alt="Screenshot 2022-02-18 at 14 28 35" src="https://user-images.githubusercontent.com/39190312/154705562-35d8b479-ea34-408f-8ada-51d80d22b2dc.png">

**Reply by sandeepdhaliwalgh** [reply]
Does the 'Add another time period' cater for 2 different shifts which start on the same day? (If that takes place).

**Reply by Sarah-Lombardi** [reply]
Yes they can add 2 shifts if they need to, in case they’ve done 2 nights in a row or something.

**Comment by tomyems** [comment]
If a user enters, 1:00 in the start time, and 2:00 in the end time. How do you decide whether they're referring to 1am to 2am, or 1pm to 2pm, or 1am to 2pm?

**Reply by Sarah-Lombardi** [reply]
It’s 24 hour clock so it would assume 1am. There’s going to be a “this is a XX hour” shift note underneath confirming how many hours it will deduct. And if they do a very very long shift and end up finishing 24 hours later, there’s a checkbox to indicate that that is the same time but the day after.

**Reply by tomyems** [reply]
How did you validate that users know its the 24 hour clock, rather than the 12 hour clock?

**Reply by Sarah-Lombardi** [reply]
Not got that far yet. Or at least they hadn’t when I moved on from the project. I’d be interested to hear how other folks have tackled that. I’m not sure how you’d do it. The good thing about the way the time card is designed is that it’s like a summary list, so once you save the time you see it in the timecard in 24 hour format. So there’s that opportunity to catch it, but would be great if it could be intelligent about it and catch mistakes before they make them.

**Comment by Dumigandesign** [comment]
Our service is a workflow tool. We are using a calendar picker that was developed by DVSA. 

This is what their developer has built: https://still-headland-16463.herokuapp.com/ 

One of our user needs was that a user may not know the exact dates but wants to search from Monday this week to Monday next week. By providing them with a date picker, they don't have to open a calendar elsewhere to check the exact dates. 

The DVSA team have had accessibility reviews with DAC and our team has done basic accessibility checks using screen readers, zoom colour contract etc. We've not had a chance to test with real user with access needs yet.

This is what it looks like on our service.

<img width="325" alt="Screenshot 2022-03-17 at 17 06 04" src="https://user-images.githubusercontent.com/83767497/158856491-735f3a83-249c-4a27-81e1-2591ba8e59f6.png">

<img width="298" alt="Screenshot 2022-03-17 at 17 06 16" src="https://user-images.githubusercontent.com/83767497/158856507-a770bac6-e3fc-415a-9251-d9f38763a5b9.png">

**Reply by LaurencePiercy** [reply]
Adding example use case here, for visibility:

"A manager wants to set up a work group where caseworkers will work on cases with the reporting date from 12 3 2022 to 12 4 2022. They would then apply that filter to the work group and all the cases that meet that filter will get allocated to the work group. 

The reason the standard GDS date input wouldn't work is that they may want to set up a date range that is from a Monday to the next week's Monday and seeing the calendar helps the user figure out which dates those fall on."

**Comment by anandamaryon-gov** [comment]
On AOT Appointment booking we're using a time input like this:
<img width="220" alt="image" src="https://user-images.githubusercontent.com/87377275/159303036-2f3777c7-69e4-4280-ac31-11b7798eef87.png">

And a variation for durations too:
<img width="214" alt="image" src="https://user-images.githubusercontent.com/87377275/159303128-29e84b85-f90f-4050-b8f5-c2bfe10f018f.png">

- We observed that our caseworking users tend to deal with 24 hour time formats, so we've used them throughout.
- We have used two inputs, for hour and minute, to simplify the validation and to align with the date input pattern.
- We show a colon between the inputs to reinforce the expected input format (HH:MM).

**Research**
We've run 4 rounds of usability testing (~5 users per round) which feature the time input as a required part of the journey. Users were allowed to choose their own times as part of the scenarios, so they weren't primed with a written time, in either 12 or 24 hour format. 

We've not observed any usability issues with the time input so far.

**Reply by tomyems** [reply]
Ah that looks good, and nice to hear about the research. Do you replay the time back to the user, say on a check your answers page? If so do you keep to the 24 hour clock?

**Reply by anandamaryon-gov** [reply]
Yes we do @tomyems 

We replay the time and date on a subsequent page like this:
![image](https://user-images.githubusercontent.com/87377275/159309613-834d421f-03b3-4e8c-a1da-bb1d60fe2475.png)

And then like this on the 'check your answers' style page:
![image](https://user-images.githubusercontent.com/87377275/159309895-c7011dc3-2272-4cd7-88e7-0cc4a771c5c9.png)

So yes, we keep it 24 hour across the service.

**Comment by jamesmckechnie** [comment]
We're recently handed over a capturing time component to devs for COP. We have not been able to test it with users yet so I have added in the design rationale behind our decisions. We intend to test the assumptions in upcoming UR. For not we've went with a 24 hour time component using two text fields and hint text.
[capturing-time-design-docs.pdf](https://github.com/UKHomeOffice/design-system/files/8362082/capturing-time-design-docs.pdf)
<img width="354" alt="Screenshot 2022-03-28 at 12 06 13" src="https://user-images.githubusercontent.com/34271054/160385302-fe312e2a-1d91-47f3-a825-2193c707682a.png">

**Comment by jbuller** [comment]
I guess you've already seen this parallel conversation, but just in case not https://github.com/alphagov/govuk-design-system-backlog/issues/173

**Comment by LaurencePiercy** [comment]
Yep, though thank you for the heads up!

We've got to do our own guidance, I think because we work with so many users who are wired into the 24 hour clock.

best

L
________________________________
